KAFKA-18135: ShareConsumer HB UnsupportedVersion msg mixed with Consumer HB

peterxcli opened the pull request:
UnsupportedVersion error is handled in the parent AbstractHeartbeatRequestManager, so used by consumer and share consumer. If a share consumer gets the errors, it will end up with a msg that is currently specific to consumer

https://github.com/apache/kafka/blob/6fd951a9c0aa773060cd6bbf8a8b8c47ee9d2965/clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractHeartbeatRequestManager.java#L384-L386

Handle the UnsupportedVersion separately in the existing 
handleSpecificError (note that the unsupported version for consumer may also end up containing a msg specific to SubscriptionPattern not supported in HB v0, if regex is used without the required v1)

- [x] add `onSpecificFailure` for consumerHB and shareHB
- [x] add `testUnsupportedVersionGeneratedOnTheBroker` to cover both handleSpecificError and handleSpecificFailure/exception

peterxcli committed:
Add specific error handling for unsupported version in share consumer and consumer

peterxcli committed:
refactor(consumer): consolidate share protocol error message and improve error handling

- Move SHARE_PROTOCOL_NOT_SUPPORTED_MSG constant to AbstractHeartbeatRequestManager
- Add test case for unsupported version error handling in ShareHeartbeatRequestManager
- Improve error message handling for regex resolution errors

AndrewJSchofield commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractHeartbeatRequestManager.java:
```diff
@@ -100,6 +100,9 @@ public abstract class AbstractHeartbeatRequestManager<R extends AbstractResponse
         "group protocol. Set group.protocol=classic on the consumer configs to revert to the CLASSIC protocol " +
         "until the cluster is upgraded.";
 
+    public static final String SHARE_PROTOCOL_NOT_SUPPORTED_MSG = "The cluster does not support the new SHARE " +
```
I would change it to "The cluster does not support the share group protocol. To use share groups, the cluster must have the share group protocol enabled." I think this message will remain valid over the next few releases without us needing to update it.

lianetm replied:
Also, I would expect this constant should go in the ShareMembershipMgr right?

lianetm commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractHeartbeatRequestManager.java:
```diff
@@ -408,8 +411,10 @@ private void onErrorResponse(final R response, final long currentTimeMs) {
+                if (!handleSpecificError(response, currentTimeMs)) {
+                    logger.error("{} failed due to {}: {}", heartbeatRequestName(), error, errorMessage);
+                    handleFatalFailure(error.exception(errorMessage));
+                }
```
I wonder if it would be simpler to remove the UnsupportedVersion handling from here completely? In the end we're duplicating the logic we already have in the `default` case further down right? We would only need the logic in the Share/Consumer managers, simply because even if the error is not specific the handling is. Thoughts?

peterxcli replied:
You're right, I think the logic in default and UnsupportedVersion handling are the same. Thanks!

lianetm commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerHeartbeatRequestManager.java:
```diff
@@ -97,6 +99,18 @@ public boolean handleSpecificError(final ConsumerGroupHeartbeatResponse response
+            case UNSUPPORTED_VERSION:
+                // Handle consumer-specific unsupported version error
+                String message = CONSUMER_PROTOCOL_NOT_SUPPORTED_MSG;
+                if (errorMessage.contains(REGEX_RESOLUTION_NOT_SUPPORTED_MSG)) {
+                    // If the error is about regex subscription, use the original error message
+                    message = errorMessage;
+                }
```
I would expect we don't need this here. This regex error is generated on the client when building the request (it generates a response with exception) so the handling lands on the `onFailure` path (not on the onResponse path that uses this `handleSpeficifError`). 

I would expect we just need to have here the same handling we had  for the UnsupportedVersion in the parent class before this PR:
https://github.com/apache/kafka/blob/ee4264439ddda7bdebcaa845752b824abba14161/clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractHeartbeatRequestManager.java#L411-L412

That being said, we do need to ensure that we handle responses for each consumer, to cover this bit that is now on the abstract mgr (so it could wrongly apply a consumer msg to the share consumer):

https://github.com/apache/kafka/blob/ee4264439ddda7bdebcaa845752b824abba14161/clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractHeartbeatRequestManager.java#L322-L325

Both consumers need to ensure they propagate their specific msgs when that exception is in a response.

peterxcli replied:
So I guess we should add a handle `onSpeficFailure` in each consumer just like `handleSpecificError` did?

lianetm replied:
Yes. The implementation of that onSpecificFailure/handleSpecificExceptionInResponse on each consumer should cover this bit I expect, with the correct msg:
https://github.com/apache/kafka/blob/497f50048305733606769a688b940cadce98521f/clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractHeartbeatRequestManager.java#L322-L325

And in the AbstractHBMgr onFailure probably just need something like: 
```
    if (exception instanceof RetriableException) {
        ...
    } else if (!onSpecificFailure()) {
        handleFatalFailure(exception);
    }
``` 
We don't need any specifics for the regex case I would expect. It should land on the handleFatalFailure and propagate the msg it has (it will only ever be generated when building consumer HB, so no risk for the share consumer)

With that seems we cover the specifics of handling responses with exceptions (generated on the client) , and in a very similar way to how we handle responses with error codes (generated on the broker).

lianetm reviewed (commented):
Thanks for the changes! Took a first look, here are some comments

peterxcli committed:
Address Andrew and Lianet's comment on SHARE_PROTOCOL_NOT_SUPPORTED_MSG

peterxcli committed:
Remove duplicated logic of unsupported version handling in AbstractHeartbeatRequestManager#onErrorResponse

AndrewJSchofield commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ShareHeartbeatRequestManager.java:
```diff
@@ -205,4 +208,25 @@ void reset() {
         }
     }
+
+    @Override
```
Please move this method earlier in the class, just after the constructors so the layout matches `ConsumerHeartbeatRequestManager`. This makes it easier for us to keep them in step over time.

AndrewJSchofield commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ShareHeartbeatRequestManager.java:
```diff
@@ -205,4 +208,25 @@ void reset() {
             }
         }
     }
+
```
Please could we have the `InheritDoc` annotation as for the other methods.

AndrewJSchofield reviewed (changes requested):
A few more changes requested. I have been working with the code on my local environment, and I think there's still something not totally right. This PR isn't making it worse in any way, but I think there's another improvement we can make here which would improve it more.

AndrewJSchofield commented:
@peterxcli It seems to me that this doesn't quite work as well as it might. In my testing, connecting a share consumer to a broker which does not support `SHARE_GROUP_HEARTBEAT` does not emit the new message. What actually happens is that `org.apache.kafka.common.errors.UnsupportedVersionException: The node does not support SHARE_GROUP_HEARTBEAT` is thrown.

The consumer checks the ApiVersions response and notices that `SHARE_GROUP_HEARTBEAT` is not supported before it's even sent a request. As a result, in `AbstractHeartbeatRequestManager.makeHeartbeatRequest`, in the `whenComplete` call, the response is null. This means we end up in `onFailure` rather than `onResponse`, and this means the new code is never actually executed.

What do you think @lianetm? Am I missing something?

AndrewJSchofield commented:
@peterxcli Actually, I've realised that this is actually a temporary problem. In AK 4.0, `SHARE_GROUP_HEARTBEAT` is marked as an unstable API. If the broker doesn't support the RPC at all, then it will behave as I said. In AK 4.1, I expect we'll mark the API as stable, and then the consumer *will* send the request. Then the new message will be used if share groups are not enabled. So, I think this is as good as it can be and the approach in the PR is OK.

lianetm commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/ShareHeartbeatRequestManagerTest.java:
```diff
@@ -422,6 +425,29 @@ public void testHeartbeatResponseOnErrorHandling(final Errors error, final boole
 
+    @ParameterizedTest
+    @ValueSource(strings = {SHARE_PROTOCOL_NOT_SUPPORTED_MSG})
+    public void testUnsupportedVersion(String errorMsg) {
```
this is truly `testUnsupportedVersionGeneratedOnTheClient` (client response with exception). Renaming will maybe clarify, and we should ensure we have coverage for something like `testUnsupportedVersionGeneratedOnTheBroker` (client response with `ErrorCode`), is it already covered?

The outcome of both tests should be the same, but they execute diff code paths modified in this PR (handleSpecificError and handleSpecificFailure/exception)

lianetm commented:
Just to answer @AndrewJSchofield's comment https://github.com/apache/kafka/pull/18101#issuecomment-2539728469 ,  you got it right, we basically have 2 paths to get the Unsupported: 
1. generated on the client when building the request if the broker does not support the API or version required => HB managers get a client response object with exception
2. generated on the broker (API supported but not enabled) => HB managers get client response with ErrorCode

For the KafkaConsumer we can have the 2 paths now but Share will always go down 1 as you described until the API becomes stable.

peterxcli committed:
Address Andrew comments on shareheartbeatrequestmanageer

peterxcli committed:
Improve error handling in HeartbeatRequestManager

- Split error handling into two distinct methods: handleSpecificFailure and
  handleSpecificExceptionInResponse
- Move UnsupportedVersion handling from onFailure to specific implementations
- Rename handleSpecificError to handleSpecificExceptionInResponse for clarity
- Improve error logging messages to better distinguish between client-side and
  broker-side version errors

This change makes the error handling more modular and clearer by separating
client-side request failures from broker response errors.

peterxcli committed:
Improve UnsupportedVersion test coverage

- Split UnsupportedVersion tests into client-side and broker-side cases
- Add separate test methods for each scenario in both Consumer and Share
  heartbeat managers
- Fix test helper methods to properly simulate client vs broker version errors
- Rename test methods to better describe test scenarios

This change improves test clarity by explicitly testing both client-side and

peterxcli commented:
Three unrelated tests failed.
UserQuotaTest | testProducerConsumerOverrideLowerQuota(String, String).quorum=kraft.groupProtocol=classic | org.opentest4j.AssertionFailedError: Client with id=QuotasTestProducer-1 should have been throttled, 0.0 ==> expected: <true> but was: <false> | 6.92s
-- | -- | -- | --
PlaintextAdminIntegrationTest | testConsumerGroups(String, String).quorum=kraft.groupProtocol=consumer | org.opentest4j.AssertionFailedError: Expected the offset for partition 0 to eventually become 1. | 21.57s
UserClientIdQuotaTest | testThrottledProducerConsumer(String, String).quorum=kraft.groupProtocol=classic | org.opentest4j.AssertionFailedError: Client with id=QuotasTestProducer-!@#$%^&*() should have been throttled, 0.0 ==> expected: <true> but was: <false> | 3.31s

AndrewJSchofield commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractHeartbeatRequestManager.java:
```diff
@@ -461,15 +434,24 @@ protected void handleFatalFailure(Throwable error) {
     }
 
+    /**
+     * Error handling specific failure to a group type when sending the request.
```
I would add "and no response has been received" to the comment. This is the crucial difference between the usage of `handleSpecificFailure` and `handleSpecificExceptionInResponse`.

AndrewJSchofield commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerHeartbeatRequestManager.java:
```diff
@@ -91,12 +94,43 @@ public ConsumerHeartbeatRequestManager(
+            String message = CONSUMER_PROTOCOL_NOT_SUPPORTED_MSG;
+            if (errorMessage.equals(REGEX_RESOLUTION_NOT_SUPPORTED_MSG)) {
+                message = REGEX_RESOLUTION_NOT_SUPPORTED_MSG;
+                logger.error("{} failed due to regex resolution not support: {}", heartbeatRequestName(), message);
```
nit: "regex resolution not supported".

AndrewJSchofield reviewed (changes requested):
A couple of final comments. This is a nice improvement.

peterxcli committed:
Address Andrew's comments

AndrewJSchofield reviewed (approved):
lgtm

AndrewJSchofield commented:
Failing test is known as flaky and tracked by https://issues.apache.org/jira/browse/KAFKA-18298.
